File: modules/lowalk-app-shield/android/src/main/java/expo/modules/lowalkappshield/WidgetSessionTimer.kt
package expo.modules.lowalkappshield

import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import kotlin.math.ceil
import kotlin.math.max

/** Active-session timer row — session start clock + compact remaining. */
object WidgetSessionTimer {
  fun formatStartLabel(startMs: Long?, timezoneId: String?): String {
    if (startMs == null) return "--:--"
    val tz = TimeZone.getTimeZone(timezoneId?.takeIf { it.isNotBlank() } ?: TimeZone.getDefault().id)
    val calendar = Calendar.getInstance(tz).apply { timeInMillis = startMs }
    return String.format(
      Locale.US,
      "%d:%02d",
      calendar.get(Calendar.HOUR_OF_DAY),
      calendar.get(Calendar.MINUTE),
    )
  }

  /** Compact H:MM remaining — e.g. 1:37, 0:41 (not mm:ss). */
  fun formatRemainingCompact(remainingMs: Long): String {
    val totalMinutes = max(0L, ceil(remainingMs / 60_000.0).toLong())
    val hours = totalMinutes / 60
    val minutes = totalMinutes % 60
    return if (hours > 0) {
      String.format(Locale.US, "%d:%02d", hours, minutes)
    } else {
      String.format(Locale.US, "0:%02d", minutes)
    }
  }

  fun resolveRemainingMs(viewModel: HeroWidgetStateEngine.ViewModel): Long {
    viewModel.timerRemainingMs?.let { return max(0L, it) }
    val endsAt = viewModel.sessionEndsAtMs ?: return 0L
    return max(0L, endsAt - System.currentTimeMillis())
  }

  fun progressPercent(viewModel: HeroWidgetStateEngine.ViewModel): Int {
    val ratio = viewModel.progressRatio ?: 0f
    return (ratio.coerceIn(0f, 1f) * 100f).toInt()
  }
}

File: lib/widgetSessionTimer.ts
/** Active-session timer row for the focus widget — start clock + compact remaining. */
import type { HeroWidgetSnapshot } from "@/types/heroWidget";

export type ActiveSessionTimer = {
  progressRatio: number;
  startLabel: string;
  remainingLabel: string;
};

export function formatSessionStartLabel(startMs: number, referenceDate = new Date(startMs)): string {
  const hours = referenceDate.getHours();
  const minutes = referenceDate.getMinutes();
  return `${hours}:${String(minutes).padStart(2, "0")}`;
}

/** Compact H:MM remaining — e.g. 1:37, 0:41 (not mm:ss). */
export function formatRemainingCompact(remainingMs: number): string {
  const totalMinutes = Math.max(0, Math.ceil(remainingMs / 60_000));
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  if (hours > 0) return `${hours}:${String(minutes).padStart(2, "0")}`;
  return `0:${String(minutes).padStart(2, "0")}`;
}

export function buildActiveSessionTimer(
  snapshot: HeroWidgetSnapshot,
  nowMs = Date.now(),
): ActiveSessionTimer | null {
  if (snapshot.state !== "active") return null;

  const startsAt = snapshot.sessionStartsAtMs;
  const endsAt = snapshot.sessionEndsAtMs;
  if (startsAt == null || endsAt == null) return null;

  const totalMs = Math.max(1, endsAt - startsAt);
  const remainingMs = Math.max(0, endsAt - nowMs);
  const progressRatio =
    snapshot.progressRatio ?? Math.min(1, Math.max(0, (totalMs - remainingMs) / totalMs));

  return {
    progressRatio,
    startLabel: formatSessionStartLabel(startsAt),
    remainingLabel: formatRemainingCompact(remainingMs),
  };
}
